### DISCUSSION TITLE
V.3 Обновление данных через pop-up в табличной части

### DISCUSSION THREAD

**Discussion by forest-lynx** [moonshine-software/moonshine#1381] [Ideas]
Доброго дня!
При обновлении поля в табличной части через модальное окно по умолчанию отображается текстовое поле, хотя данный метод доступен и для остальных видов полей. Например для поля выбора (Select) так же будет выведено поле в виде обычного текстового поля, а не поля выбора как ожидалось. В методе __invoke() классf UpdateOnPreviewPopover имеем: 
```php
//...
->fields([
    Flex::make([
        //...
        Text::make('Title', 'value')
            ->style('margin: 0!important')
            ->setValue($this->field->toFormattedValue())
            ->withoutWrapper(),
    ]),
])
//...
```

**Предложение:** 
Подставлять вместо обычного текстового поля само поле которое редактируется, например "клонированное" через рефлексию, для примера метод по клонированию: 
```php
function cloneField(FieldContract $field): FieldContract {
      $reflection = new ReflectionClass($field);
      $clone = $reflection->newInstance();

      $properties = $reflection->getProperties();
      foreach ($properties as $property) {
          $property->setAccessible(true);
          $propertyName = $property->getName();

          $value = match($propertyName) {
              'updateOnPreview',
              'updateOnPreviewPopover' => false,
              'updateOnPreviewParentComponent',
              'updateOnPreviewUrl',
              'onChangeUrl' => null,
              'attributes' => $property->getValue($field)->except([
                  '@change',
                  'data-async-events',
                  'data-async-method',
              ]),
              default => $property->getValue($field)
          };

          $property->setValue($clone, $value);
      }

      $clone->customAttributes([
          'name' => $clone->getNameAttribute(),
      ])->withoutWrapper();

      return $clone;
  }
```
В классе UpdateColumnFormRequest можем получить правила валидации для данного поля из ресурса.

ЗЫ: Кроме того, логичнее было бы значение для кнопки задать не фиксированное а с учетом локали.

**Comment by lee-to** [comment]
Это полноценное issue а не дискуссия, там тестовая портянка осталась

**Comment by lee-to** [comment]
Вот фикс - https://github.com/moonshine-software/moonshine/pull/1383
Будет круто если протестишь у себя

**Reply by forest-lynx** [reply]
Ок погляжу. А как насчет валидации.

**Reply by forest-lynx** [reply]
Посмотрел вроде как все работает. Единственное по поводу валидации еще можно подтянуть из ресурса.

**Reply by lee-to** [reply]
Что имеешь ввиду?

**Reply by forest-lynx** [reply]
В методе rules() UpdateColumnFormRequest например можно сделать так:
```php
//...
'value' => [
    'present',
    ...$this->getResource()->getRules()[request()->getScalar('field')]
]
//...
```

**Reply by lee-to** [reply]
Понял, попробую добавить

**Reply by lee-to** [reply]
Проверяй! Добавил

**Reply by forest-lynx** [reply]
Пробежался по-быстрому, вроде как всё работает.
